base/src/commonMain/kotlin/utils/BufferedAsListChannel.kt

```diff
@@ -0,0 +1,96 @@
+ *     You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ */
+
+package com.splendo.kaluga.base.utils
+
+import com.splendo.kaluga.base.singleThreadDispatcher
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ClosedReceiveChannelException
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.yield
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A Rendezvous Channel that buffers all elements sent to it in a list until the next receive
+ * @param T the type of element to batch. Must be non-nullable
+ */
+interface BufferedAsListChannel<T : Any> : SendChannel<T>, ReceiveChannel<List<T>>
+
+/**
+ * Creates a [BufferedAsListChannel] that batches its elements with a given [CoroutineContext]
+ * @param T the type of element to batch. Must be non-nullable
+ * @param coroutineContext the [CoroutineContext] to use for batching
+ * @return the [BufferedAsListChannel] created
+ */
+fun <T : Any> BufferedAsListChannel(coroutineContext: CoroutineContext): BufferedAsListChannel<T> = BufferedAsListChannelInt(coroutineContext)
+
+internal class BufferedAsListChannelInt<T : Any> private constructor(
+    private val sendChannel: Channel<T>,
+    private val receiveChannel: Channel<List<T>>,
+    coroutineContext: CoroutineContext
+) : BufferedAsListChannel<T>, SendChannel<T> by sendChannel, ReceiveChannel<List<T>> by receiveChannel {
+
+    constructor(
+        coroutineContext: CoroutineContext
+    ) : this(
+        Channel<T>(Channel.UNLIMITED),
+        Channel<List<T>>(),
+        coroutineContext
+    )
+
+    init {
+        val dispatcher = singleThreadDispatcher("GroupingChannel")
+        CoroutineScope(coroutineContext + dispatcher).launch {
+            while (!sendChannel.isClosedForReceive && !receiveChannel.isClosedForSend) {
+                try {
+                    val list = mutableListOf(sendChannel.receive())
+                    do {
+                        do {
+                            val lastReceived = sendChannel.tryReceive().getOrNull()?.also {
+                                list.add(it)
+                            }
+                        } while (lastReceived != null)
+                    } while (
+                        !receiveChannel.isClosedForSend && receiveChannel.trySend(list).isFailure.also { yield() }
+                    )
```

Comment:
This and the surrounding part could use comments as I have to work backwards what the intention is.

There is still a potential loop with just a `yield` but I guess `receiveChannel.trySend(list)` failing. This is right now a `Channel<List<T>>()` with no parameters, so `RENDEZVOUS` channel. So this will loop (at max CPU) until the channel is read. 

As the `ReceiveChannel` tends to be busier than the `SendChannel` we would expect this to occur normally. 

I there a reason not to use `send` instead of `trySend`? (or use a `select` over the condition in the `while`).

Sure, we will not take more of the receive channel but this has an infinite buffer that will build up while the receiver is busy, and when it's done they're all put in a list together. This, I believe is more or less the intention.

Reply:
the proper way is probably selectors over both the channels, e.g. because you want to cancel both if one of them closes etc.

Reply:
I went with a select approach instead. Reason to use trySend instead of onSend was to prevent missing events that happened between the moment the sendChannel became temporarily empty and waiting for the receive channel to send.

Select solves this in a suspended manner so its the proper approach. See latest commit